### PR TITLE
Quote v2: add transforms

### PR DESCRIPTION
**Pull request by oandregal**
Part of https://github.com/WordPress/gutenberg/issues/15486
Related https://github.com/WordPress/gutenberg/pull/25892

## What?

This PR adds transforms to the Quote v2.

## How?

Transform | Before | After |
--- | --- | --- |
From paragraph | :heavy_check_mark:  | :heavy_check_mark: Removed. Superseeded by the `*` transform. |
From heading | :heavy_check_mark:  | :heavy_check_mark: Removed. Superseeded by the `*` transform. |
From Pullquote | :heavy_check_mark:  | :heavy_check_mark: Same behaviour. It's only available if all inner blocks are paragraphs. | 
From `>` prefix | :heavy_check_mark: | :yellow_circle: See TODO section. Focus stays in the wrapper. Before, the paragraph was focused and users could start typing. To be addressed in a [follow-up](https://github.com/WordPress/gutenberg/pull/39718#issuecomment-1079196437). |
From raw (paste) | :heavy_check_mark: Supports `<blockquote><p /><cite /></blockquote>` | :yellow_circle: See TODO section. Supports the old markup and also `<figure><blockquote/><figcaption /></figure>`. It ignores the cite from the old markup. |
From Group | - | :heavy_check_mark: |
From any (`*`) | - | :heavy_check_mark: Allow any block to be transformed into a quote |
To Paragraph | :heavy_check_mark:  | :heavy_check_mark: Removed. Superseeded by the `*` transform. |
To Heading | :heavy_check_mark:  | :heavy_check_mark: Removed. Superseeded by the `*` transform. |
To Pullquote | :heavy_check_mark:  | :heavy_check_mark: |
To Group | - | :heavy_check_mark: |
To any (`*`) | - | :heavy_check_mark: Allow a quote to be "Unwrapped" (see "Transforms > Unwrap") |

## Testing Instructions

- Go to "Gutenberg > Experiments" page and enable the Quote v2 block.
- Test that the transforms listed above work as expected.

<details>
<summary>Content for testing the `raw` transform</summary>

```html
<p>First example:</p>

<figure>
  <blockquote>
    <p>Paragraph.</p>
    <ul>
      <li>list</li>
    </ul>
    <code>rm -rf</code>
  </blockquote>
  <figcaption>Author</figcaption>
</figure>

<p>Second example:</p>

<blockquote>
  <p>Paragraph.</p>
  <ul>
    <li>list</li>
  </ul>
  <code>rm -rf</code>
</blockquote>
<figcaption>Author</figcaption>

<p>Third example:</p>

<blockquote><p>Paragraph.</p><cite>Cite</cite></blockquote>
```
</details>

## TODO

- ~Merge https://github.com/WordPress/gutenberg/pull/39729 (it updates how the block is registered and fixes the issues with `attribution`).~
- ~Update transforms to consider other attributes (anchor, align, etc). Specially when coming from pullquote.~
- For the `>` prefix transformation, investigate if we can make the focus to start in the inner content instead of  in the wrapper. Before, the paragraph was focused and users could start typing. To be addressed in a [follow-up](https://github.com/WordPress/gutenberg/pull/39718#issuecomment-1079196437).
- For the `raw` transform:
  - `attribution` from `figcaption` not working.
  - when the pasted code contains an `` element, it doesn't transfom it to a quote

### CODE REVIEW

**Comment by ellatrix** [comment]
@mtias What should happen to the pull quote block? Should that we a style variation of the quote block? Or why is that block still around?

**Comment by ellatrix** [comment]
@oandregal Are e2e tests ran with the new quote block or when do we make that switch? It would be good to know what tests are passing and failing

**Comment by ellatrix** [comment]
Two issues I'm seeing:

When I create a quote with `> `, focus doesn't move to the paragraph.

The styling of the quote looks weird. There's two lines. And when the quote is no longer selected, the lines disappear.
quote-styling

**Comment by ellatrix** [comment]
Another issue: why is there an "add attribution" below the quote and above the quote? We should pick one. When I click on the toolbar button, I can no longer select the block wrapper for some reason...
add-attribution

**Comment by ellatrix** [comment]
Hm, the raw transform is also not working for me 🤔

File: packages/block-library/src/quote/v2/transforms.js
Comment (ellatrix) [inline packages/block-library/src/quote/v2/transforms.js]: I'm not sure if it's important to look for the cite in pasted content. It could be bad if there are multiple cites or if the cite is actually part of the content. We were using cite wrongly before, so I would omit cite handling here.

Comment (oandregal) [reply packages/block-library/src/quote/v2/transforms.js]: This is only to provide backwards-compatibility with the existing `raw` behavior, that would handle this nicely:

```html
<p>content before</p>
<blockquote><p>First paragraph.</p><p>Second paragraph.</p><cite>Author</cite></blockquote>
<p>content after</p>
```

If we don't do this, things that worked before would no longer work.

**Comment by oandregal** [comment]
@ellatrix I'd like to keep this PR focused on the transforms so PR's size & concern is manageable :) The things I've seen that we need to do outside the transforms are listed [here](https://github.com/WordPress/gutenberg/pull/25892#issuecomment-1076686005) at the moment: interaction of the attribution button, deprecations, etc.

>Are e2e tests ran with the new quote block or when do we make that switch? It would be good to know what tests are passing and failing

The tests would be enabled when https://github.com/WordPress/gutenberg/pull/25892 is updated to make v2 the default. I can do that after this lands.

**Comment by oandregal** [comment]
I've just rebased to get https://github.com/WordPress/gutenberg/pull/39729 in which fixes issues with filling the `attribution` attribute.

**Comment by oandregal** [comment]
> Hm, the raw transform is also not working for me thinking

@ellatrix What html input are you using?

**Comment by oandregal** [comment]
ok, the only issue I see now is the focus on the `>` prefix transform, the rest is fixed.

File: packages/block-library/src/quote/v2/transforms.js
Comment (ellatrix) [reply packages/block-library/src/quote/v2/transforms.js]: I think that's ok, we don't need back compatibility for paste. This markup would be very rare outside of WP. This action could be more destructive if it is NOT used as a caption. See https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-cite-element.

Comment (oandregal) [reply packages/block-library/src/quote/v2/transforms.js]: OK, I'll go with your instinct. Reverted at https://github.com/WordPress/gutenberg/pull/39718/commits/9756835ec82bbaf41be12e9b13d1af7cb1c9f773

**Comment by oandregal** [comment]
Thoughts for the prefix transform `>` and the focus: we have the same issues for any prefix transform that creates content with inner blocks, as the focus will remain in the wrapper. For example, the list block v2. The prefix transforms are the kind of flow that is keyboard-driven and which would be affected by this focus issue, so I'd think this is something we want to address.

- In the `/` inserter we solved by passing a flag to the Redux action that creates the block (`updateSelection`). I've looked at this from the perspective of transforms, but I don't see an immediate way to do this. It seems to go against the current, as the return type of a transform is a block object and doesn't contain any metadata for the redux action that uses them.

- Another potential direction to investigate: make the focus go to the inner block by default. I think at the moment all actions (`INSERT_BLOCKS`, `REPLACE_BLOCKS`, `REPLACE_INNER_BLOCKS`) do the opposite by default. Originally, at least the `REPLACE_INNER_BLOCKS` had this behaviour by default but was changed in WordPress 5.7 https://github.com/WordPress/gutenberg/pull/26312 Perhaps we could consider changing it. It'd also require modifying `REPLACE_BLOCKS` to modify the `updateSelection`.

Given all this, I'd suggest we investigate this issue separately and do not block this PR by this issue.

@ellatrix thoughts?

**Comment by oandregal** [comment]
The `attribution` is now working for me consistently in the `raw` transform with this code:

```html
<figure>
  <blockquote>
    <p>Paragraph.</p>
    <ul>
      <li>list</li>
    </ul>
    <code>rm -rf</code>
  </blockquote>
  <figcaption>Author</figcaption>
</figure>
```

It doesn't with the same code but without the `figure`.

**Comment by ellatrix** [comment]
I'm still seeing the paste issue when copying the content here: https://codepen.io/iseulde/pen/eYyvJbd

There's no quote and the two paragraphs in the quote disappear.

When I remove the image, it works as expected indeed. Let's try to figure this out separately then.

**Review by ellatrix — APPROVED** [review]
Let's merge 😄

**Pull request merged**
